cranelift/codegen/src/isa/x64/inst/emit.rs

```diff
@@ -0,0 +1,2987 @@
+    match size {
+        8 | 4 => sink.put4(simm32),
+        2 => sink.put2(simm32 as u16),
+        1 => sink.put1(simm32 as u8),
+        _ => panic!("x64::Inst::emit_simm: unreachable"),
+    }
+}
+// The top-level emit function.
+//
+// Important!  Do not add improved (shortened) encoding cases to existing
+// instructions without also adding tests for those improved encodings.  That
+// is a dangerous game that leads to hard-to-track-down errors in the emitted
+// code.
+//
+// For all instructions, make sure to have test coverage for all of the
+// following situations.  Do this by creating the cross product resulting from
+// applying the following rules to each operand:
+//
+// (1) for any insn that mentions a register: one test using a register from
+//     the group [rax, rcx, rdx, rbx, rsp, rbp, rsi, rdi] and a second one
+//     using a register from the group [r8, r9, r10, r11, r12, r13, r14, r15].
+//     This helps detect incorrect REX prefix construction.
+//
+// (2) for any insn that mentions a byte register: one test for each of the
+//     four encoding groups [al, cl, dl, bl], [spl, bpl, sil, dil],
+//     [r8b .. r11b] and [r12b .. r15b].  This checks that
+//     apparently-redundant REX prefixes are retained when required.
+//
+// (3) for any insn that contains an immediate field, check the following
+//     cases: field is zero, field is in simm8 range (-128 .. 127), field is
+//     in simm32 range (-0x8000_0000 .. 0x7FFF_FFFF).  This is because some
+//     instructions that require a 32-bit immediate have a short-form encoding
+//     when the imm is in simm8 range.
+//
+// Rules (1), (2) and (3) don't apply for registers within address expressions
+// (|Addr|s).  Those are already pretty well tested, and the registers in them
+// don't have any effect on the containing instruction (apart from possibly
+// require REX prefix bits).
+//
+// When choosing registers for a test, avoid using registers with the same
+// offset within a given group.  For example, don't use rax and r8, since they
+// both have the lowest 3 bits as 000, and so the test won't detect errors
+// where those 3-bit register sub-fields are confused by the emitter.  Instead
+// use (eg) rax (lo3 = 000) and r9 (lo3 = 001).  Similarly, don't use (eg) cl
+// and bpl since they have the same offset in their group; use instead (eg) cl
+// and sil.
+//
+// For all instructions, also add a test that uses only low-half registers
+// (rax .. rdi, xmm0 .. xmm7) etc, so as to check that any redundant REX
+// prefixes are correctly omitted.  This low-half restriction must apply to
+// _all_ registers in the insn, even those in address expressions.
+//
+// Following these rules creates large numbers of test cases, but it's the
+// only way to make the emitter reliable.
+//
+// Known possible improvements:
+//
+// * there's a shorter encoding for shl/shr/sar by a 1-bit immediate.  (Do we
+//   care?)
+
```

Comment:
```suggestion

/// The top-level emit function.
///
/// Important!  Do not add improved (shortened) encoding cases to existing
/// instructions without also adding tests for those improved encodings.  That
/// is a dangerous game that leads to hard-to-track-down errors in the emitted
/// code.
///
/// For all instructions, make sure to have test coverage for all of the
/// following situations.  Do this by creating the cross product resulting from
/// applying the following rules to each operand:
///
/// (1) for any insn that mentions a register: one test using a register from
///     the group [rax, rcx, rdx, rbx, rsp, rbp, rsi, rdi] and a second one
///     using a register from the group [r8, r9, r10, r11, r12, r13, r14, r15].
///     This helps detect incorrect REX prefix construction.
///
/// (2) for any insn that mentions a byte register: one test for each of the
///     four encoding groups [al, cl, dl, bl], [spl, bpl, sil, dil],
///     [r8b .. r11b] and [r12b .. r15b].  This checks that
///     apparently-redundant REX prefixes are retained when required.
///
/// (3) for any insn that contains an immediate field, check the following
///     cases: field is zero, field is in simm8 range (-128 .. 127), field is
///     in simm32 range (-0x8000_0000 .. 0x7FFF_FFFF).  This is because some
///     instructions that require a 32-bit immediate have a short-form encoding
///     when the imm is in simm8 range.
///
/// Rules (1), (2) and (3) don't apply for registers within address expressions
/// (`Addr`s).  Those are already pretty well tested, and the registers in them
/// don't have any effect on the containing instruction (apart from possibly
/// require REX prefix bits).
///
/// When choosing registers for a test, avoid using registers with the same
/// offset within a given group.  For example, don't use rax and r8, since they
/// both have the lowest 3 bits as 000, and so the test won't detect errors
/// where those 3-bit register sub-fields are confused by the emitter.  Instead
/// use (eg) rax (lo3 = 000) and r9 (lo3 = 001).  Similarly, don't use (eg) cl
/// and bpl since they have the same offset in their group; use instead (eg) cl
/// and sil.
///
/// For all instructions, also add a test that uses only low-half registers
/// (rax .. rdi, xmm0 .. xmm7) etc, so as to check that any redundant REX
/// prefixes are correctly omitted.  This low-half restriction must apply to
/// _all_ registers in the insn, even those in address expressions.
///
/// Following these rules creates large numbers of test cases, but it's the
/// only way to make the emitter reliable.
///
/// Known possible improvements:
///
/// * there's a shorter encoding for shl/shr/sar by a 1-bit immediate.  (Do we
///   care?)
```